Add torch.library.opcheck

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124496

This PR:
- exposes torch.testing._internal.optests.opcheck as
  torch.library.opcheck
- Adds support for CustomOpDef (aka functions decorated with
  torch.library.custom_op) to opcheck.

Test Plan:
- Updated tests
- We validated opcheck's design internally.

zou3519 committed:
Add torch.library.opcheck

This PR:
- exposes torch.testing._internal.optests.opcheck as
  torch.library.opcheck
- Adds support for CustomOpDef (aka functions decorated with
  torch.library.custom_op) to opcheck.

Test Plan:
- Updated tests
- We validated opcheck's design internally.

[ghstack-poisoned]

albanD commented on docs/source/library.rst:
```diff
@@ -4,13 +4,23 @@ torch.library
+Testing custom ops
+------------------
+
+Use :func:`torch.library.opcheck` to test custom ops for incorrect usage of the
```
Why not `torch.testing.custom_op.opcheck` ?

zou3519 replied:
It's probably too many namespaces for people to keep track of? The analogy is torch.autograd.gradcheck -- would we have wanted to put it into torch.testing.autograd.gradcheck?

albanD replied:
torch.testing didn't exist when gradcheck was added?
I'm happy to open the discussion to move gradcheck there if you think that we should do a larger effort.

zou3519 replied:
Resolved offline: we'll go with torch.library.opcheck for now, but we're happy to put it into torch.testing too

zou3519 committed:
Update on "Add torch.library.opcheck"


This PR:
- exposes torch.testing._internal.optests.opcheck as
  torch.library.opcheck
- Adds support for CustomOpDef (aka functions decorated with
  torch.library.custom_op) to opcheck.

Test Plan:
- Updated tests
- We validated opcheck's design internally.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
